File: leetcode/92. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        left -= 1
        right -= 1
        vhead = ListNode()
        vhead.next = head
        p1 = vhead
        
        idx = 0
        while idx < left:
            p1 = p1.next
            idx += 1
        
        left = p1.next
        p2 = left
        while idx < right:
            p2 = p2.next
            idx += 1
        right = p2
        p2 = p2.next

        x = left
        y = left.next
        x.next = None
        
        while y != p2:
            x, y.next, y = y, x, y.next
        
        p1.next = right
        left.next = p2

        return vhead.next
        

File: leetcode/92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode dummy;
        dummy.next = head;

        ListNode* p = head;
        ListNode* p0 = &dummy;
        int n = 1;
        while (n != left) {
            n += 1;
            p = p->next;
            p0 = p0->next;
        }

        if (p == nullptr || p->next == nullptr) {
            return dummy.next;
        }

        ListNode* first = p;
        ListNode* p1 = p->next;
        ListNode* p2 = p->next;

        while (n != right) {
            n += 1;
            p2 = p2->next;
            p1->next = p;
            p = p1;
            p1 = p2;
        }

        first->next = p2;
        p0->next = p;
        return dummy.next;
    }
};
